### ISSUE TITLE
backupd: possible to index data that's not on disk

### ISSUE THREAD

**Issue by elliefm** [cyrusimap/cyrus-imapd#1814]
After a couple of long backup replications were abruptly stopped (one by ^C'ing sync_client, one by killing backupd), verify on the affected backup found this:

```
checking chunk 3 checksums...
 checking file checksum...
 checking data length
data length mismatch for chunk 3: 6301696000 on disk,6301697343 in index
error
```

`backup_append()` doesn't call `backup_index()` until it is sure the data is written, so it's sureness is clearly suspect.  Note that backupd calls `backup_append()` with the flush argument true, so we _are_ calling `gzflush()` with Z_FULL_FLUSH here.  https://refspecs.linuxbase.org/LSB_3.0.0/LSB-PDA/LSB-PDA/zlib-gzflush-1.html

**Comment by elliefm** [comment]
A later compact run fell apart on the truncated data too, not good!

```
splitting chunk 3
splitting chunk 3
splitting chunk 3

didn't parse dlist, error -1
error reading chunk at offset 223912403, byte 2006728704: Input/output error
compact [snip]: failed (System I/O error)
```

**Comment by elliefm** [comment]
Regarding the original issue, I think the problem may be that `backupd_reset()` doesn't properly close any backups open at the time that it's closed, so a clean shutdown isn't so clean.  Need to experiment.

**Comment by elliefm** [comment]
Fixed in 44ce5fd5
